haystack/core/super_component/utils.py

```diff
@@ -53,34 +59,83 @@ def _types_are_compatible(type1: T, type2: T) -> bool:
+            return True, cast(Optional[T], result_type)
+        return False, None
 
-
-def _check_non_union_compatibility(type1: T, type2: T, type1_origin: Any, type2_origin: Any) -> bool:
+    if type2_origin is Union and type1_origin is not Union:
+        # Find all compatible types from the union
+        compatible_types = []
+        for union_arg in get_args(type2):
+            is_compat, common = _types_are_compatible(type1, union_arg)
+            if is_compat and common is not None:
+                compatible_types.append(common)
+        if compatible_types:
+            # The constructed Union or single type must be cast to Optional[T]
+            # to satisfy mypy, as T is specific to this function's call context.
+            result_type = Union[tuple(compatible_types)] if len(compatible_types) > 1 else compatible_types[0]
+            return True, cast(Optional[T], result_type)
+        return False, None
+
+    # Both are Union types
+    compatible_types = []
+    for arg1 in get_args(type1):
+        for arg2 in get_args(type2):
+            is_compat, common = _types_are_compatible(arg1, arg2)
+            if is_compat and common is not None:
+                compatible_types.append(common)
+
+    if compatible_types:
+        # The constructed Union or single type must be cast to Optional[T]
+        # to satisfy mypy, as T is specific to this function's call context.
+        result_type = Union[tuple(compatible_types)] if len(compatible_types) > 1 else compatible_types[0]
+        return True, cast(Optional[T], result_type)
+    return False, None
+
+
+def _check_non_union_compatibility(
+    type1: T, type2: T, type1_origin: Any, type2_origin: Any
+) -> Tuple[bool, Optional[T]]:
     """Handle non-Union type compatibility cases."""
     # If no origin, compare types directly
     if not type1_origin and not type2_origin:
-        return type1 == type2
+        if type1 == type2:
+            return True, type1
+        return False, None
 
     # Both must have origins and they must be equal
     if not (type1_origin and type2_origin and type1_origin == type2_origin):
-        return False
+        return False, None
 
     # Compare generic type arguments
     type1_args = get_args(type1)
     type2_args = get_args(type2)
 
     if len(type1_args) != len(type2_args):
-        return False
-
-    return all(_types_are_compatible(t1_arg, t2_arg) for t1_arg, t2_arg in zip(type1_args, type2_args))
+        return False, None
+
+    # Check if all arguments are compatible
+    common_args = []
+    for t1_arg, t2_arg in zip(type1_args, type2_args):
+        is_compat, common = _types_are_compatible(t1_arg, t2_arg)
+        if not is_compat:
+            return False, None
+        common_args.append(common)
+
+    # Reconstruct the type with common arguments
+    return True, cast(Optional[T], type1_origin[tuple(common_args)])
```

Comment:
Would it be possible here to somehow reuse the typing library types instead of the python builtin ones? Right now we are trying to consistently using typing types and will switch over to python builtin ones later this year. And I believe by calling `type1_origin` like this is what causes `List[int]` to become `list[int]`